### PR TITLE
Adds metrics for timing of loading accounts

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

There are not *fine-grained* metrics on how long it takes to load accounts.

#### Summary of Changes

Add a histogram for how long it takes to load accounts, and submit those metrics.

Since this can be intrusive, the histogram and metrics are behind a (rust) feature. To enable, build with `--features load-accounts-historgram`. For example:

```
$ cargo build --release --bin solana-validator --features load-accounts-histogram
```

Here's an example of what these metrics look like:

Screenshot 2023-09-12 at 12 23 51 PM

### CODE REVIEW

File: accounts-db/src/accounts_db/load_accounts_histogram.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    }
+
+    /// Submits datapoint
+    pub fn submit(&mut self) {
+        let duration = self.previous_submit.elapsed(); // do this before building the histogram
+        let samples = std::mem::take(&mut self.samples); // use `take` because we want to clear `samples`
+        let histogram = Self::build_histogram(&samples);
+
+        datapoint_info!(
```
Comment (apfitzge) [inline accounts-db/src/accounts_db/load_accounts_histogram.rs:43]: I'd assume it wouldn't happen, but if there are no samples do we still want to report this?

File: accounts-db/src/accounts_db/load_accounts_histogram.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            samples: Vec::new(),
+            previous_submit: Instant::now(),
+        }
+    }
+
+    /// Records `sample`
+    ///
+    /// Will be included in the next submission
+    pub fn record(&mut self, sample: Duration) {
+        self.samples.push(sample);
```
Comment (apfitzge) [inline accounts-db/src/accounts_db/load_accounts_histogram.rs:27]: We add a sample everytime we load an account. Should we put a bound on number of samples so this doesn't grow too large?
I think reasonably it wouldn't, but potentially, although probably very unlikely, there's a resource exhaustion attack here with a spammer sending txs loading many accounts that fail quickly so banking stage is able to process a lot of them?

**Review by apfitzge — COMMENTED** [review]
Since the metrics are behind a cargo feature, I'm wondering who they are for. Is this just something we'll be running internally to gather metrics? I'd imagine not many validators will run the feature if it's not default.

File: Cargo.toml [hunk 3]
```diff
@@ -204,6 +204,7 @@ generic-array = { version = "0.14.7", default-features = false }
 gethostname = "0.2.3"
 getrandom = "0.2.10"
 goauth = "0.13.1"
+hdrhistogram = "7.5.2"
```
Comment (sakridge) [inline Cargo.toml:207]: There's already a histogram crate, what is the difference/benefit of adding another one?

Comment (apfitzge) [reply Cargo.toml:207]: > Should I look into bumping our usage of the histogram crate from v0.6 to v0.7?

Seems like it should be separate from this PR, but if it has features that improve the interface, I vote yes 😄

**Comment by apfitzge** [comment]
> And yes, I wouldn't want validators to run this by default, since it is somewhat invasive.

Appreciate the detailed response. Adding more detailed and intrusive metrics doesn't seem that uncommon - I know I've definitely kept branches around to do similar things.

There's probably some consideration on whether we should be having a more encompassing feature, such as `gather_intrusive_metrics` which enables all these intrusive metrics, or leave it granular with the user being responsible for enabling each detailed metric feature.
I'm a bit on the fence, but I think the case is niche enough that we can leave it granular.

Potentially, we could have both w/ granular and a global feature, something like
`#[cfg(any(granular_metric_feature, gather_intrusive_metrics))`. Wdyt?

**Comment by sakridge** [comment]
Generally we prefer runtime flags and not features. Is there a strong argument for the feature? seems like the additional code and runtime check is probably not hurting performance much?

**Pull request closed without merging**